cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java

```diff
@@ -0,0 +1,179 @@
+
+import io.ballerina.cli.cmd.RunCommand;
+import io.ballerina.projects.ProjectKind;
+import io.ballerina.projects.util.FileUtils;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * Represents a watcher that looks into project file changes.
+ *
+ * @since 2201.10.0
+ */
+public class ProjectWatcher {
+    private static final String BAL_EXTENSION = ".bal";
+    private static final String TOML_EXTENSION = ".toml";
+    private static final String DEPENDENCIES_TOML = "Dependencies.toml";
+    private static final String RESOURCES_DIR = "resources";
+    private static final String MODULES_DIR = "modules";
+
+    private final WatchService fileWatcher;
+    private final Map<WatchKey, Path> watchKeys;
+    private final RunCommand runCommand;
+    private final PrintStream outStream;
+    private final Path projectPath;
+    private final ProjectKind projectKind;
+
+    public ProjectWatcher(RunCommand runCommand, Path projectPath, PrintStream outStream) throws IOException {
+        this.fileWatcher = FileSystems.getDefault().newWatchService();
+        this.runCommand = runCommand;
+        this.projectPath = projectPath.toAbsolutePath();
+        this.outStream = outStream;
+        this.watchKeys = new HashMap<>();
+        this.projectKind = deriveProjectKind();
+        registerFileTree(projectPath);
+    }
+
+    public void watch() throws IOException, InterruptedException {
+        RunCommandExecutor thread = new RunCommandExecutor(runCommand, outStream);
+        thread.start();
+        while (thread.shouldWatch()) {
+            WatchKey key;
+            key = fileWatcher.poll();
+            Path dir = watchKeys.get(key);
+            if (dir == null) {
+                continue;
+            }
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind<?> kind = event.kind();
+                if (kind == OVERFLOW) {
+                    continue;
+                }
+                WatchEvent<Path> pathWatchEvent = cast(event);
+                Path changedFileName = pathWatchEvent.context();
+                Path changedFilePath = dir.resolve(changedFileName).toAbsolutePath();
+                if (isValidFileChange(changedFilePath)) {
+                    outStream.println("\\nDetected file changes. Re-running the project...");
```

Comment:
This prints `\nDetected file changes. Re-running the project...`. Need to remove the leading \n

Reply:
Resolved with 3d007562f6199380baff4b3b43bb29cbd1b7709c